includes/api/ApiQueryAllUsers.php

```diff
@@ -0,0 +1,383 @@
+			$useIndex = false;
+			// Filter only users don't belong to a given group
+			$this->addTables( 'user_groups', 'ug1' );
+
+			if ( count( $params['excludegroup'] ) == 1 ) {
+				$exclude = array( 'ug1.ug_group' => $params['excludegroup'][0] );
+			} else {
+				$exclude = array( $db->makeList( array( 'ug1.ug_group' => $params['excludegroup'] ), LIST_OR ) );
+			}
+			$this->addJoinConds( array( 'ug1' => array( 'LEFT OUTER JOIN',
+				array_merge( array( 'ug1.ug_user=user_id' ), $exclude )
+				)
+			) );
+			$this->addWhere( 'ug1.ug_user IS NULL' );
+		}
+
+		if ( $params['witheditsonly'] ) {
+			$this->addWhere( 'user_editcount > 0' );
+		}
+
+		$this->showHiddenUsersAddBlockInfo( $fld_blockinfo );
+
+		if ( $fld_groups || $fld_rights ) {
+			// Show the groups the given users belong to
+			// request more than needed to avoid not getting all rows that belong to one user
+			$groupCount = count( User::getAllGroups() );
+			$sqlLimit = $limit + $groupCount + 1;
+
+			$this->addTables( 'user_groups', 'ug2' );
+			$this->addJoinConds( array( 'ug2' => array( 'LEFT JOIN', 'ug2.ug_user=user_id' ) ) );
+			$this->addFields( 'ug2.ug_group ug_group2' );
+		} else {
+			$sqlLimit = $limit + 1;
+		}
+
+		if ( $params['activeusers'] ) {
+			global $wgActiveUserDays;
+			$this->addTables( 'recentchanges' );
+
+			$this->addJoinConds( array( 'recentchanges' => array(
+				'INNER JOIN', 'rc_user_text=user_name'
+			) ) );
+
+			$this->addFields( 'COUNT(*) AS recentedits' );
+
+			$this->addWhere( "rc_log_type IS NULL OR rc_log_type != 'newusers'" );
+			$timestamp = $db->timestamp( wfTimestamp( TS_UNIX ) - $wgActiveUserDays*24*3600 );
+			$this->addWhere( "rc_timestamp >= {$db->addQuotes( $timestamp )}" );
+
+			$this->addOption( 'GROUP BY', $userFieldToSort );
+		}
+
+		$this->addOption( 'LIMIT', $sqlLimit );
+
+		$this->addFields( array(
+			'user_name',
+			'user_id'
+		) );
+		$this->addFieldsIf( 'user_editcount', $fld_editcount );
+		$this->addFieldsIf( 'user_registration', $fld_registration );
+
+		if ( $useIndex ) {
+			$this->addOption( 'USE INDEX', array( 'user' => 'user_name' ) );
+		}
+
+		$res = $this->select( __METHOD__ );
+
+		$count = 0;
+		$lastUserData = false;
+		$lastUser = false;
+		$result = $this->getResult();
+
+		//
+		// This loop keeps track of the last entry.
+		// For each new row, if the new row is for different user then the last, the last entry is added to results.
+		// Otherwise, the group of the new row is appended to the last entry.
+		// The setContinue... is more complex because of this, and takes into account the higher sql limit
+		// to make sure all rows that belong to the same user are received.
+
+		foreach ( $res as $row ) {
+			$count++;
+
+			if ( $lastUser !== $row->user_name ) {
+				// Save the last pass's user data
+				if ( is_array( $lastUserData ) ) {
+					$fit = $result->addValue( array( 'query', $this->getModuleName() ),
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "query" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)